### ISSUE TITLE
Replace pycrypto with pycryptodome

### ISSUE THREAD

**Issue by Rovanion** [pycontribs/jira#619]
When using oauth this python package uses the pycrypto library. The problem is that pycrypto  stopped being developed four years ago and has not been touched since. This entails a number of issues, probably including security issues, but what got me started down this path was the dependency hell it landed me in. 

Since pycrypto is four years old it depends on an even older version of cryptography, and long story short this led me into pinning old versions of two other libraries than pycrypto that pycrypto depended on. 

But the story doesn't end there, which probably is a good thing considering for how long pycrypto has gone unmaintained. The software I'm building is kept in a docker container during development, a container running Alpine Linux. Alpine has switched from OpenSSL to LibreSSL and in the end it was impossible to both satisfy the old requirements of pycrypto and those of postgresql.

The solution to this issue is fortunately quite straight forward: Switching to a maintained fork named pycryptodome. 

In my limited testing this library works as a drop in replacement for pycrypto without any changes to the jira python library.

For further reference here is an issue against PyCrypto which has been linked by many others in the same situation as you: https://github.com/dlitz/pycrypto/issues/238

**Comment by ssbarnea** [comment]
@Rovanion I don't mind doing a switch to a maintained library but I see few problems:
- new library is hosted under individual github account, which is single-point of failure kind of issue.
- i do not see build status icons on the main page
- apparently this fork is not published on pypi - or if it is, is not documented in the readme
- concerns regarding availability of precompiled versions on major linux distros: deb and rpm based ones.

Switching a dependency can be tricky if you are not switching it with the right thing. Lets gather some extra info.

**Comment by Rovanion** [comment]
Yes absolutely. I'm starting to think that we might have read different pages since we're seeing such different information. But here is the information you requested:

1.  new library is hosted under individual github account, which is single-point of failure kind of issue.

I see your point. But on the other hand a single point of failure is better than something which has already failed.

2.  i do not see build status icons on the main page

They're linked in the readme: https://github.com/Legrandin/pycryptodome  
https://travis-ci.org/Legrandin/pycryptodome  
https://ci.appveyor.com/project/Legrandin/pycryptodome  

3.  apparently this fork is not published on pypi - or if it is, is not documented in the readme

The readme says that it comes in two versions through pypi:  
https://pypi.org/project/pycryptodome/  
https://pypi.org/project/pycryptodomex/

4.  concerns regarding availability of precompiled versions on major linux distros: deb and rpm based ones.

It seems to already be in Ubuntu 18.04:   
https://packages.ubuntu.com/search?keywords=pycryptodome&searchon=names&suite=bionic&section=all

5. also, make a PR and let's see how it looks.

I'm not quite sure how to approach this though. Pycrypto seems to only be mentioned in the documentation of this project. Though that's just based off of a search through the code here on github. Are there any tests which tests the oauth code paths?

**Comment by ssbarnea** [comment]
If this is the case, the fix for this would be to bump the minimal version of oauthlib so we do not endup getting pycrypto. 

Make a PR for this but be sure you don't ask for a too recent version. Also include a comment explaining why the bump was needed, like `lib>=1.23  # to avoid pycrypto dependency`

**Comment by dchevell** [comment]
This library doesn't have a direct dependency on oauthlib, that comes from requests-oauthlib which still specifies `oauthlib[signedtoken]>=0.6.2`
It's simple enough to add a direct dependency here:
```
...
defusedxml
pbr>=3.0.0
oauthlib[signedtoken]>=1.0.0
requests-oauthlib>=0.6.1
...
```
Given that the removal of pycrypto happened in 1.0.0, that should be enough to ensure safety.

Since I don't have an appropriate test case for this I'll leave it to @Rovanion!

**Comment by Rovanion** [comment]
Doing such a modification to pycontribs/jira seems to have successfully solved the issue. I opened up a PR upstreaming this modification with #627.
